### PR TITLE
Add a filter for maintainers in the jobset-eval view

### PR DESCRIPTION
**Pull request by Ma27**
###### Summary

Continuation of my efforts from #685.

During the release-managers' talk [from last NixCon](https://www.youtube.com/watch?v=gYQSrv4KP_Q&feature=youtu.be&list=TLPQMjcwNDIwMjAsfqiZkeibjw&t=992) it has been suggested to add a search-feature to list all failing packages a user maintains. Theoretically this is already possible by introspecting the job-list in the dashboard of a Hydra-user, but not everyone owns a google-account (which is usually needed for the upstream-Hydra), also not everyone uses their google-email for NixOS stuff (I don't, for instance).

To summarize, the following things changed:

* Instead of saving all maintainers' emails as a comma-separated string in the builds table (and therefore creating a lot of duplications), all maintainers are stored in a separate table and are associated to a build using a many-to-many relationship in DBIx. Whenever maintainers are needed in a controller-action, those can be fetched using an additional JOIN. For the REST-API, the maintainers are fetched by default when requesting details of a build.

  This has the additional benefit that if the GitHub-handle of a user changes, Hydra updates the existing maintainer entry.
* The filter-form in the jobset-eval view now contains an additional select-field to choose whether to filter by name or by maintainer (for backwards-compatibility, this is optional and by default it's filtered by job-name).

For a detailed overview, I suggest to review the commits including their messages.

###### Screenshots

* The jobset-eval overview has been slightly changed to display and filter maintainers:
2020-04-28-000224_screenshot

* The detail-view of a build now displays the github-handles by default:
2020-04-27-233855_screenshot

* If a maintainer doesn't have a github-handle, the email is shown as fallback:
2020-04-27-234030_screenshot

###### Pending work

* [ ] Provide a way to upgrade. The migration I hacked together just drops the `maintainers` field from all entries in the `builds`-table which is bad if you need to search older jobsets or want to receive notifications if a job is restarted. I had the following ideas (though at least @grahamc might have more experience here):
     * add a migration-approach as implemented for the `jobset_id`-fixes
     * re-add the old `maintainers`-column and merge the data from this column with the data from the `maintainer`-table
* [ ] Investigate if there's a way to retrieve the symbols of a maintainer (e.g. `lib.maintainers.ma27`). In that case it's even possible for Hydra to keep track of changes to the GitHub handle and email-address of a maintainer.
* [ ] (bonus-points) add support for this in [hydra-check](https://github.com/nix-community/hydra-check)

-----

cc @edolstra @grahamc

### CODE REVIEW

**Comment by gilligan** [comment]
Are you still working on this? 

I guess you still need help on the database migration matter? How about the part with retrieving maintainer info?

**Comment by Ma27** [comment]
> Are you still working on this?

Yes.

> I guess you still need help on the database migration matter?

Well, the question is much rather, how do we want to migrate? Should we still keep the `maintainers`-column for backwards-compatibility or should we do some sort of `hydra-backfill-maintainers`-thing just like it happened with the `jobset_id` migration?

Or do you have an even better idea? :) 

> How about the part with retrieving maintainer info?

I'm not even sure if that's possible tbh. Does anybody have anough experience with the C++ API to accomplish that? (or, as a dirty hack, one may fetch the key from `lib.maintainers` if `email`/`github` match, however this doesn't work if a 3rdparty-project has its own list of maintainers).

**Comment by Ma27** [comment]
@gilligan what do you think of my suggestions regarding the DB-migration? And who else should I also ping for a review?

While I'm happy to continue working on this, I'd prefer to get some feedback on this first :)

**Comment by grahamc** [comment]
@Ma27 do you have a few hundred gigabytes of free space to try out the migration on the prod dataset? If so, I can get you a copy.

**Comment by Ma27** [comment]
@grahamc,

> do you have a few hundred gigabytes of free space

will be a bit close, but should be possible :) 

----

But first of all we should agree on how to implement the migration (I posted two ideas in the original comment).

**Comment by Ma27** [comment]
Just for the record, I'm happy to continue working on this (including a rebase to master :p), but first of all, I'd like to get some feedback and to get the migration-topic resolved.

**Comment by Ma27** [comment]
@grahamc in case you have time, could you please take a look at my comments about the migration needed for this change?

**Comment by Ma27** [comment]
ping @grahamc @gilligan

**Comment by Ma27** [comment]
ping @grahamc how should we proceed with the migration question?

**Comment by Ma27** [comment]
@grahamc Thanks a lot for the detailed feedback!

> For the migration, in my most humble opinion it would be really good to retain the historical data. I think we might not have chosen to if we wrote the code ourselves for hydra.nixos.org, but I think overall other users of Hydra would find it important to have the data preserved.

Agreed. That's why I brought this up :) 

> Do you have any opinions or ideas on this? I have a preference for the simplicity of the second one, but also am not so keen on the maintainer data being NULL.

I agree. Rather than having to keep the upgrade-path code in `nixpkgs` (https://github.com/NixOS/nixpkgs/pull/83600/) and to add more complexity to this change, it's preferable to let users wait for a short amount of time until the migration is fully finished IMHO. In the end, this isn't a mission-critical thing anyway from my PoV.

> I'd also want to see the backfiller tested thoroughly, since that final drop would be inconvenient if it did the wrong thing.

This is actually the part I'm mostly anxious about. As soon as there's a working draft for `hydra-backfill-ids` we could test this with live data from `hydra.nixos.org` for instance as you suggested a while ago.

File: src/sql/hydra.sql [hunk 1]
```diff
@@ -216,6 +215,21 @@ create table Builds (
...
+);
+
+create table BuildsByMaintainers (
+    maintainer_id   integer not null,
+    build_id        integer not null,
+
+    primary key (maintainer_id, build_id),
+    foreign key (maintainer_id) references Maintainers(id),
+    foreign key (build_id) references Builds(id)
+);
```
[the remark this answers is not in the document]
Comment (Ma27) [reply src/sql/hydra.sql:232]: > I was thinking we should just store the list of maintainers per build directly in the build table

That's what we're currently doing. This has the downside that it's harder to query maintainers and to keep track of e.g. changed github handles.

File: src/root/common.tt [hunk 2]
```diff
@@ -139,6 +140,15 @@ BLOCK renderBuildListBody;
...
       [% END %]
+      <td>
+        [% UNLESS build.maintainers.size %]
+          <em>not given</em>
+        [% ELSE %]
+          [% FOREACH m IN build.maintainers %]
+            <em>[% maintainer(m) %]</em>[% UNLESS loop.last %], [% END %]
+          [% END %]
+        [% END %]
+      </td>
```
[the remark this answers is not in the document]
Comment (Ma27) [reply src/root/common.tt:151]: > Which page is this rendered on?

* Jobset evals (this are probably the biggest issue if `?full=1` is set).
* Search page
* Status page (running builds etc)
* "My builds" (if your email matches the one from your Hydra account you can already see maintained derivation-builds).

> I'm a bit unsure about how often we want to query and display this data, in particular I'm thinking tables which are very heavy with many rows of builds.

While the biggest problem will probably be the jobset-eval page, this is the one where I really want to have this feature. I suggest to test this with live-data and try to optimize the query (and caching) if necessary.

File: src/lib/Hydra/Schema/Builds.pm [hunk 3]
```diff
@@ -634,6 +652,7 @@ my %hint = (
         buildoutputs => 'name',
         buildproducts => 'productnr',
         buildmetrics => 'name',
+        maintainers => 'github_handle',
```
[the remark this answers is not in the document]
Comment (Ma27) [reply src/lib/Hydra/Schema/Builds.pm:655]: Fair point. Will take care of this.

File: src/script/hydra-eval-jobset [hunk 4]
```diff
@@ -468,6 +467,28 @@ sub checkBuild {
             , ischannel => $buildInfo->{isChannel}
             });
 
+        for my $maintainer (@{$buildInfo->{maintainers}}) {
+            my $entity = $db->resultset('Maintainer')
+                ->search({ email => $maintainer->{email} })
+                ->single;
```
[the remark this answers is not in the document]
Comment (Ma27) [reply src/script/hydra-eval-jobset:473]: Frankly, I'm not sure since I don't know *this* ORM implementation. I'll make sure that some sort of caching mechanism is involved.

**Comment by Ma27** [comment]
Btw I'll try to implement the migration during the next weeks (as soon as I have time to) as we seem to agree to go for the second option. As soon as this is done, I'll try to do some live-testing to make sure that this doesn't explode with real-world data.

**Comment by siraben** [comment]
Any updates on this? It would be nice to have and closes #101741

**Comment by Ma27** [comment]
Just left a comment in the issue you've linked: I don't think that it should be blocked by this. When it's ready we can still adapt the script.

Regarding the updates: it's still missing a sane migration path and I'm sorry, but I didn't have the time yet to work on that. Hopefully, I'll be able to finish that in December :)

**Comment by Ma27** [comment]
Just pushed a small batch of changes related to the DB migration. Currently, the following things are tbd to consider this ready:

* `hydra-update-maintainers` should perform DB operations in batches rather than everything at once.
* I need to address the comments from @grahamc 
* @grahamc btw, how should we test a migration with productive data? E.g. via a subset of Hydra's live data?

**Comment by Ma27** [comment]
@grahamc would you mind taking a look at the migration so we can make a plan on how to test it? When I have more time & motivation for this again, I'd take care of your review comments from July :)

File: src/script/hydra-eval-jobset (hunk 4 above)
[the remark this answers is not in the document]
Comment (Ma27) [reply src/script/hydra-eval-jobset]: Implemented a simple hash-based cache.

File: src/lib/Hydra/Schema/Builds.pm (hunk 3 above)
[the remark this answers is not in the document]
Comment (Ma27) [reply src/lib/Hydra/Schema/Builds.pm]: Done.

**Comment by Ma27** [comment]
@edolstra would you mind leaving a review here? :)

**Comment by Ma27** [comment]
cc @NixOS/hydra-triage any chance to get a review here? :)

File: src/lib/Hydra/Schema/Buildsbymaintainer.pm [hunk 5]
```diff
@@ -0,0 +1,111 @@
+use utf8;
+package Hydra::Schema::Buildsbymaintainer;
```
Comment (grahamc) [inline src/lib/Hydra/Schema/Buildsbymaintainer.pm:2]: Can you add `"buildsbymaintainer" => "BuildsByMaintainer",` to `src/sql/updated-dbix.pl`'s moniker_map?

Comment (Ma27) [reply src/lib/Hydra/Schema/Buildsbymaintainer.pm:2]: @grahamc fixed :)

**Comment by Ma27** [comment]
@grahamc did you have a chance to take a look? :)

**Comment by Ma27** [comment]
@grahamc @edolstra rebased onto latest `master`. Would be cool if you could take a look :)

**Comment by Ma27** [comment]
@grahamc may I ask if you'd have some time for a review / test of this? I guess I may have some more timeslots to in the near future to e.g. take care of the next rebase and respond to comments / change requests :)

**Comment by Ma27** [comment]
This could use tests for at least the channel output and Email notifications btw. But the current testcase should cover most of the stuff that's relevant already :)

**Comment by Ma27** [comment]
@cole-h @grahamc merged latest master into this :)

**Comment by Ma27** [comment]
@cole-h @grahamc @edolstra any chance to get a review? Would really like to proceed with this :)

**Comment by Ma27** [comment]
@cole-h @grahamc do you currently have the capacity to move this forward?

**Comment by Ma27** [comment]
Since I was rebasing other branches anyways, I decided to just update this one as well :) 

cc @grahamc @cole-h if you'd like to review this and need some assistance, let me know!

Note: even if we're finding the time to get this ready, I'd suggest to not use it for this ZHF period since it requires some database migration tasks that I'd rather not do at the time of the year, where the hydra.nixos.org has its most users :)

**Comment by Ma27** [comment]
@grahamc should we try to get this into a mergable state at some point now? %)

**Comment by Ma27** [comment]
@grahamc, so I just did the following things (as we discussed.... quite a while ago):

* Merged master into the branch and ensured that my VM test for the schema update works.
* Tested the database upgrade locally with the stuff from https://hydra.ist.nicht-so.sexy/ (it's more than it looks because I'm using the private-projects feature from #919 there ;-) ) and it worked flawlessly.

IIRC the next step is to test this somewhere against hno, correct? In that case I'd probably need a test-instance since I'm pretty sure that none of my servers has sufficient disk-space (though I don't know the exact numbers, but I'd expect the db to be >100G at least, right)?

**Comment by Ma27** [comment]
I don't see myself working on this anymore. ALso, there's other tooling like zh.fail doing the same.
